carcara/src/ast/pool/mod.rs

```diff
@@ -184,6 +237,54 @@ impl PrimitivePool {
                 result.push(self.compute_sort(body));
                 Sort::Function(result)
             }
+            Term::IndexedOp { op, op_args, args } => {
+                let sort = match op {
+                    IndexedOperator::BvExtract => {
+                        /*
+                        ! extract has restrictions regarding their op_args
+                        ((_ extract i j) (_ BitVec m) (_ BitVec n))
+
+                        where
+                        - i, j, m, n are numerals
+                        - m > i ≥ j ≥ 0,
+                        - n = i - j + 1
+                         */
```

Comment:
@bpandreotti I remember we talked about asserting argument values of indexed operators following SMT-LIB rules.

I did something like this on line [264](https://github.com/ufmg-smite/carcara/pull/22/files#diff-ed7104440530d2d2bc9fc8116dce3419f215a94325ac670b9f99c821072b2a1bR264), but ideally I should return an error.

The problem is that this function only returns an Rc<Term> and I'd have to change it's return type to also include errors, and therefore this would impact other places.

However, this is the smaller of problems. My point is: shouldn't argument values be checked during the parser run? Specifically at the same moment we check the amount of arguments of each Term.

Reply:
PS: This is not valid for concat operator specifically, but for every other BV operator. Just used concat as an example.

Reply:
> However, this is the smaller of problems. My point is: shouldn't argument values be checked during the parser run? Specifically at the same moment we check the amount of arguments of each Term.

I probably misspoke before, but you're absolutely right, these restrictions on the arguments should be checked when parsing the term (in `Parser::make_indexed_op` to be precise). Here, we can just assume that the term is well-sorted, that is, that the arguments follow the restrictions.